src/serac/infrastructure/profiling.hpp

```diff
@@ -112,22 +119,49 @@ auto&& forwarder(T&& thing)
   return std::forward<T>(thing);
 }
 
-}  // namespace detail
+/**
+ * @brief Guarantees str is a c string
+ */
+inline const char* make_cstr(const char* str) { return str; }
+
+/**
+ * @brief Converts a std::string into a c string
+ */
+inline const char* make_cstr(const std::string& str) { return str.c_str(); }
+
+}  // namespace serac::profiling::detail
 /// @endcond
 
+#define SERAC_PROFILE_SCOPE(name) \
+  cali::ScopeAnnotation SERAC_CONCAT(region, __LINE__)(serac::profiling::detail::make_cstr(name))
+
 /**
  * @brief The type that should be returned from the profiling wrapper lambda
  * It converts rvalue reference arguments of type T&& to T to avoid returning
  * a reference to a temporary and leaves all other types intact.
  */
-#define SERAC_PROFILE_EXPR_RETURN_TYPE(expr) detail::remove_rvalue_reference<decltype(detail::forwarder(expr))>::type
+#define SERAC_PROFILE_EXPR_RETURN_TYPE(expr) \
+  serac::profiling::detail::remove_rvalue_reference<decltype(serac::profiling::detail::forwarder(expr))>::type
 
-#define SERAC_PROFILE_EXPR(name, expr)                                \
-  [&]() -> SERAC_PROFILE_EXPR_RETURN_TYPE(expr) {                     \
-    const cali::ScopeAnnotation SERAC_CONCAT(region, __LINE__)(name); \
-    return static_cast<SERAC_PROFILE_EXPR_RETURN_TYPE(expr)>(expr);   \
+#define SERAC_PROFILE_EXPR(name, expr)                                                                     \
+  [&]() -> typename SERAC_PROFILE_EXPR_RETURN_TYPE(expr) {                                                 \
+    const cali::ScopeAnnotation SERAC_CONCAT(region, __LINE__)(serac::profiling::detail::make_cstr(name)); \
+    return static_cast<typename SERAC_PROFILE_EXPR_RETURN_TYPE(expr)>(expr);                               \
   }()
 
+/**
+ * @brief Profiles an expression several times; Return the last evaluation
+ */
+#define SERAC_PROFILE_EXPR_LOOP(name, expr, ntest)                                                                  \
```

Comment:
In this case I think it could be pretty easy for the compiler to recognize that the loop has no side effects, in which case it could optimize it out.

Google Benchmark implements a [DoNotOptimize](https://github.com/google/benchmark/blob/main/include/benchmark/benchmark.h#L330-L366) function that avoids this - possibly we should look into this.